proposals/p0821.md

```diff
@@ -0,0 +1,1119 @@
+separate layer between the pointer object and the pointee object, and introduce
+dereference syntax (both the unary `*` operator and the `->` operator).
+
+C++ makes an important extension to this model to represent _smart pointers_ by
+allowing the dereference operators to be overloaded. This can be seen across a
+wide range of APIs such as `std::unique_ptr`, `std::shared_ptr`,
+`std::weak_ptr`, etc. These user-defined types preserve a fundamental property
+of C++ pointers: the separation between the pointer object and the pointee
+object.
+
+The distinction between pointer and pointee is made syntactically explicit in
+C++ both when _dereferencing_ a pointer, and when _forming_ the pointer or
+taking an object's address. These two sides can be best illustrated when
+pointers are used for function parameters. The caller code must explicitly take
+the address of an object to pass it to the function, and the callee code must
+explicitly dereference the pointer to access the caller-provided object.
+
+### References
+
+C++ provides for indirection _without_ the syntactic separation of pointers:
+references. Because a reference provides no syntactic distinction between the
+reference and the referenced object--that is their point!--it is impossible to
+refer to the reference itself in C++. This creates a number of restrictions on
+their design:
+
+-   They _must_ be initialized when declared
+-   They cannot be rebound or unbound.
+-   Their address cannot be taken.
+
+References were introduced originally to enable operator overloading, but have
+been extended repeatedly and as a consequence fill a wide range of use cases.
+Separating these and understanding them is essential to forming a cohesive
+proposal for Carbon -- that is the focus of the rest of our analysis of
+references here.
+
+#### Special but critical case of `const T&`
+
+One form of reference in C++ has unique properties: `const T&` for some type
+`T`, or a _`const` reference_. The primary use for these is also the one that
+motivates its unique properties: a zero-copy way to provide an input function
+parameter without requiring the syntactic distinction in the caller and callee
+needed when using a pointer. The intent is to safely emulate passing by-value
+without the cost of copying. Provided the usage is immutable, this emulation can
+safely be done with a reference and so a `const` reference fits the bill here.
+
+However, to make zero-copy, pass-by-value to work in practice, it must be
+possible to pass a _temporary_ object. That works well with by-value parameters
+after all. To make this work, C++ allows a `const` reference to bind to a
+temporary. However, the rules for parameters and locals are the same in C++ and
+so this would create serious lifetime bugs. This is fixed in C++ by applying
+_lifetime extension_ to the temporary. The result is that `const` references are
+quite different from other references, but they are also quite useful: they are
```

Comment:
I don't understand why `const&` is "quite different from other references" - all declarations of reference type do lifetime extension of prvalues, excepting non-const lvalue since that's an error.

```
int&& x = 0;
```

also does lifetime extension.